File: src/main/java/com/qtech/bubbles/effect/PoisonEffect.java
package com.qtech.bubbles.effect;

import com.jhlabs.image.HSBAdjustFilter;
import com.qtech.bubbles.common.effect.StatusEffect;
import com.qtech.bubbles.common.effect.StatusEffectInstance;
import com.qtech.bubbles.entity.Entity;
import com.qtech.bubbles.entity.damage.DamageSource;
import com.qtech.bubbles.entity.damage.DamageSourceType;
import com.qtech.bubbles.event.FilterEvent;
import com.qtech.utilities.python.builtins.ValueError;
import org.bson.BsonDocument;
import org.bson.BsonInt64;

public class PoisonEffect extends StatusEffect {
    public PoisonEffect() throws ValueError {
        super();
    }

    @Override
    public void onFilter(StatusEffectInstance statusEffectInstance, FilterEvent evt) {
        HSBAdjustFilter filter = new HSBAdjustFilter();
        filter.setHFactor((float) (System.currentTimeMillis() - statusEffectInstance.getStartTime()) / 3000 % 1);
        evt.addFilter(filter);
    }

    @Override
    protected boolean canExecute(Entity entity, StatusEffectInstance statusEffectInstance) {
        return System.currentTimeMillis() >= statusEffectInstance.getTag().getInt64("nextDamage").getValue();
    }

    @Override
    public void execute(Entity entity, StatusEffectInstance statusEffectInstance) {
        entity.getGameType().attack(entity, (double) statusEffectInstance.getStrength() / 2, new DamageSource(null, DamageSourceType.POISON));
        BsonDocument tag = statusEffectInstance.getTag();
        BsonInt64 nextDamage = tag.getInt64("nextDamage");
        tag.put("nextDamage", new BsonInt64(nextDamage.getValue() + 2000));
    }

    @Override
    public void onStart(StatusEffectInstance statusEffectInstance, Entity entity) {
        BsonDocument tag = statusEffectInstance.getTag();
        tag.put("nextDamage", new BsonInt64(System.currentTimeMillis() + 2000));
        tag.put("startTime", new BsonInt64(System.currentTimeMillis()));
    }

    @SuppressWarnings("EmptyMethod")
    @Override
    public void onStop(Entity entity) {
        // Do nothing
    }

    @Override
    protected void updateStrength() {
        // Do nothing
    }
}

File: src/main/kotlin/com/qtech/bubbles/effect/PoisonEffect.kt
package com.qtech.bubbles.effect

import com.jhlabs.image.HSBAdjustFilter
import com.qtech.bubbles.common.effect.Effect
import com.qtech.bubbles.common.effect.EffectInstance
import com.qtech.bubbles.common.entity.DamageSourceType
import com.qtech.bubbles.entity.Entity
import com.qtech.bubbles.common.entity.EntityDamageSource
import com.qtech.bubbles.event.FilterEvent

class PoisonEffect : Effect() {
    override fun onFilter(effectInstance: EffectInstance, evt: FilterEvent) {
        val filter = HSBAdjustFilter()
        filter.setHFactor((System.currentTimeMillis() - effectInstance.startTime).toFloat() / 3000 % 1)
        evt.addFilter(filter)
    }

    override fun canExecute(entity: Entity, effectInstance: EffectInstance): Boolean {
        return System.currentTimeMillis() >= effectInstance.tag.getLong("nextDamage")
    }

    override fun execute(entity: Entity, effectInstance: EffectInstance) {
        entity.gameMode.attack(entity, effectInstance.strength.toDouble() / 2, EntityDamageSource(null, DamageSourceType.POISON))
        val tag = effectInstance.tag
        val nextDamage = tag.getLong("nextDamage")
        tag.putLong("nextDamage", nextDamage + 2000)
    }

    override fun onStart(effectInstance: EffectInstance, entity: Entity) {
        val tag = effectInstance.tag
        tag.putLong("nextDamage", System.currentTimeMillis() + 2000)
        tag.putLong("startTime", System.currentTimeMillis())
    }

    override fun onStop(entity: Entity?) {
        // Do nothing
    }

    override fun updateStrength() {
        // Do nothing
    }
}